src/v/cluster/tm_stm_cache.h

```diff
@@ -280,63 +235,41 @@ class tm_stm_cache {
         return ids;
     }
 
-    fragmented_vector<tm_transaction> get_all_transactions() {
-        fragmented_vector<tm_transaction> ans;
-        if (_mem_term) {
-            auto entry_it = _state.find(_mem_term.value());
-            if (entry_it != _state.end()) {
-                for (const auto& [_, tx] : entry_it->second.txes) {
-                    ans.push_back(tx);
-                }
-                for (const auto& [id, tx] : _log_txes) {
-                    auto tx_it = entry_it->second.txes.find(id);
-                    if (tx_it == entry_it->second.txes.end()) {
-                        ans.push_back(tx);
-                    }
-                }
-                return ans;
-            }
-        }
-        return get_log_transactions();
-    }
+    fragmented_vector<tm_transaction> get_all_transactions();
 
-    std::deque<tm_transaction> checkpoint() {
-        std::deque<tm_transaction> txes_to_checkpoint;
+    std::deque<tm_transaction> checkpoint();
 
-        if (_mem_term == std::nullopt) {
-            return txes_to_checkpoint;
-        }
-        auto term = _mem_term.value();
+    std::optional<tm_transaction> oldest_tx();
 
-        auto entry_it = _state.find(term);
-        if (entry_it == _state.end()) {
-            return txes_to_checkpoint;
-        }
-        auto& entry = entry_it->second;
+    size_t tx_cache_size();
 
-        auto can_transfer = [](const tm_transaction& tx) {
-            return !tx.transferring
-                   && (tx.status == tm_transaction::ready || tx.status == tm_transaction::ongoing);
-        };
-        // Loop through all ongoing/pending txns in memory and checkpoint.
+    ~tm_stm_cache() noexcept { clear_log(); }
 
-        for (auto& [_, tx] : entry.txes) {
-            if (can_transfer(tx)) {
-                txes_to_checkpoint.push_back(tx);
-            }
-        }
+private:
+    struct tx_wrapper {
+        tx_wrapper() {}
 
-        return txes_to_checkpoint;
-    }
+        tx_wrapper(const tm_transaction& tx) { this->tx = tx; }
+
+        tm_transaction tx;
+        safe_intrusive_list_hook _hook;
+    };
+
+    // Tracks the LRU order of tx sessions. When the count exceeds
+    // max_transactions_per_coordinator, we abort tx sessions in the
+    // LRU order.
+    counted_intrusive_list<tx_wrapper, &tx_wrapper::_hook> lru_txes;
```

Comment:
it isn't about size, it's about keeping the list of tx sorted with O(1) instead of O(n ln n).

Reply:
right, my question was why counted_intrusive_list and not intrusive_list with auto unlinking. counted variant requires explicit unlinking (hence a footgun) and here we don't even need const time size because we can get it from _log_txes.